### ISSUE TITLE
Add an option to specify ldd executable

### ISSUE THREAD

**Issue by nurupo** [probonopd/linuxdeployqt#295]
If my system is `x86_64-linux-gnu system` and I have cross-compiled my Qt app to `arm-linux-gnueabi` using the , running `ldd` on my app won't produce any meaningful output:

```
$ ldd foo
        not a dynamic executable
```

However, there are ways to get around that, e.g. the [`xldd`](https://github.com/crosstool-ng/crosstool-ng/blob/master/scripts/xldd.in) from crosstool-ng project:

```
$ xldd foo
        libQt5Network.so.5 => /usr/lib/arm-linux-gnueabi/libQt5Network.so.5 (0xdeadbeef)
        libQt5Core.so.5 => /usr/lib/arm-linux-gnueabi/libQt5Core.so.5 (0xdeadbeef)
        libpthread.so.0 => /lib/arm-linux-gnueabi/libpthread.so.0 (0xdeadbeef)
        libc.so.6 => /lib/arm-linux-gnueabi/libc.so.6 (0xdeadbeef)
        ld-linux.so.3 => /lib/ld-linux.so.3 (0xdeadbeef)
        libz.so.1 => /lib/arm-linux-gnueabi/libz.so.1 (0xdeadbeef)
        libicui18n.so.57 => /usr/lib/arm-linux-gnueabi/libicui18n.so.57 (0xdeadbeef)
        libicuuc.so.57 => /usr/lib/arm-linux-gnueabi/libicuuc.so.57 (0xdeadbeef)
        libicudata.so.57 => /usr/lib/arm-linux-gnueabi/libicudata.so.57 (0xdeadbeef)
        libdl.so.2 => /lib/arm-linux-gnueabi/libdl.so.2 (0xdeadbeef)
        libstdc++.so.6 => /usr/lib/arm-linux-gnueabi/libstdc++.so.6 (0xdeadbeef)
        libm.so.6 => /lib/arm-linux-gnueabi/libm.so.6 (0xdeadbeef)
        libgcc_s.so.1 => /lib/arm-linux-gnueabi/libgcc_s.so.1 (0xdeadbeef)
        libpcre16.so.3 => /usr/lib/arm-linux-gnueabi/libpcre16.so.3 (0xdeadbeef)
        libdouble-conversion.so.1 => /usr/lib/arm-linux-gnueabi/libdouble-conversion.so.1 (0xdeadbeef)
        libglib-2.0.so.0 => /lib/arm-linux-gnueabi/libglib-2.0.so.0 (0xdeadbeef)
        libpcre.so.3 => /lib/arm-linux-gnueabi/libpcre.so.3 (0xdeadbeef)
        librt.so.1 => /lib/arm-linux-gnueabi/librt.so.1 (0xdeadbeef)
        libproxy.so.1 => /usr/lib/arm-linux-gnueabi/libproxy.so.1 (0xdeadbeef)
        libQt5Widgets.so.5 => /usr/lib/arm-linux-gnueabi/libQt5Widgets.so.5 (0xdeadbeef)
        libQt5Gui.so.5 => /usr/lib/arm-linux-gnueabi/libQt5Gui.so.5 (0xdeadbeef)
        libharfbuzz.so.0 => /usr/lib/arm-linux-gnueabi/libharfbuzz.so.0 (0xdeadbeef)
        libfreetype.so.6 => /usr/lib/arm-linux-gnueabi/libfreetype.so.6 (0xdeadbeef)
        libpng16.so.16 => /usr/lib/arm-linux-gnueabi/libpng16.so.16 (0xdeadbeef)
        libgraphite2.so.3 => /usr/lib/arm-linux-gnueabi/libgraphite2.so.3 (0xdeadbeef)
        libGLESv2.so.2 => /usr/lib/arm-linux-gnueabi/libGLESv2.so.2 (0xdeadbeef)
        libdrm.so.2 => /usr/lib/arm-linux-gnueabi/libdrm.so.2 (0xdeadbeef)
        libglapi.so.0 => /usr/lib/arm-linux-gnueabi/libglapi.so.0 (0xdeadbeef)
```

The library path is correct, but the address is a made up one. I don't think that linuxdeployqt uses the address in any way, so it should be fine with the output of `xldd`. However, linuxdeployqt is hardcoded to use `ldd`, there is no way to ask it to use, e.g. `/usr/bin/arm-linux-gnueabi-ldd` or anything like that.

**Comment by probonopd** [comment]
Hold on, are you saying there is a bash script that can stand in for `ldd` and works across all architectures? In this case, we might well bundle it with linuxdeployqt and use it _always_? If so, that's probably what I'd do...

**Comment by TheAssassin** [comment]
readelf can do the same:

```
> readelf -d /usr/lib/appimagelauncher/libappimageupdate.so 

Dynamic section at offset 0x4cc20 contains 29 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000e (SONAME)             Library soname: [libappimageupdate.so]
[...]
```

Basically, the information is contained in a section in the ELF header `.dynsym`. If you have a tool that understands the ELF format and can fetch section offset and length, and you know the format of this section, it isn't very hard to parse it. I would rather write some non-subprocess-dependent solution. Is there anything like a libreadelf or some other library that works cross platform and eliminates the need for any kind of subprocess management?

**Comment by nurupo** [comment]
Readelf works too, it's also [packaged](https://packages.debian.org/search?searchon=contents&keywords=readelf&mode=path&suite=stable&arch=any) in binutils of crosscompilers, so there is no need for a 3rd part ldd script. It's just that it doesn't provide the path to the library, only a library name, which probably wouldn't be enough for linuxdeployqt.

**Comment by TheAssassin** [comment]
@nurupo replicating the library searching shouldn't be too hard. You can use e.g., `ldconfig -v` as a replacement.

**Comment by nurupo** [comment]
@TheAssassin 
The xldd scans 

>Hold on, are you saying there is a bash script that can stand in for ldd and works across all architectures?

I think so, yes. So far it seems to work great for me. I have built my binary as x86_64. armel and arm64, and in both cases it found the right Qt libraries. It seems to do library search for libraries with matching bitness (elf format) and architecture, e.g. here is how it decided that my x86_64 Qt app needs a x86_64 libQt5Network.so over armel and arm64 libQt5Network.so which I also have installed (look for "skip incompatible"):

```sh
$ CT_XLDD_VERBOSE=1 /usr/bin/x86_64-linux-gnu-ldd --root / deploy/usr/bin/my-qt-app
Adding basic lib dirs
-> adding search dir '/lib'
-> adding search dir '/usr/lib'
Done adding basic lib dirs
Scanning '/etc/ld.so.conf'
    Parsing ld.so.conf: '//etc/ld.so.conf'
    -> handling include directive '//etc/ld.so.conf.d/*.conf'
        Parsing ld.so.conf: '//etc/ld.so.conf.d/aarch64-linux-gnu.conf'
        -> adding search dir '/lib/aarch64-linux-gnu'
        -> adding search dir '/usr/lib/aarch64-linux-gnu'
        Finished parsing ld.so.conf: '//etc/ld.so.conf.d/aarch64-linux-gnu.conf'
        Parsing ld.so.conf: '//etc/ld.so.conf.d/arm-linux-gnueabi.conf'
        -> adding search dir '/lib/arm-linux-gnueabi'
        -> adding search dir '/usr/lib/arm-linux-gnueabi'
        Finished parsing ld.so.conf: '//etc/ld.so.conf.d/arm-linux-gnueabi.conf'
        Parsing ld.so.conf: '//etc/ld.so.conf.d/libc.conf'
        -> adding search dir '/usr/local/lib'
        Finished parsing ld.so.conf: '//etc/ld.so.conf.d/libc.conf'
        Parsing ld.so.conf: '//etc/ld.so.conf.d/x86_64-linux-gnu.conf'
        -> adding search dir '/lib/x86_64-linux-gnu'
        -> adding search dir '/usr/lib/x86_64-linux-gnu'
        Finished parsing ld.so.conf: '//etc/ld.so.conf.d/x86_64-linux-gnu.conf'
        Parsing ld.so.conf: '//etc/ld.so.conf.d/zz_i386-biarch-compat.conf'
        -> adding search dir '/lib32'
        -> adding search dir '/usr/lib32'
        Finished parsing ld.so.conf: '//etc/ld.so.conf.d/zz_i386-biarch-compat.conf'
        Parsing ld.so.conf: '//etc/ld.so.conf.d/zz_x32-biarch-compat.conf'
        -> adding search dir '/libx32'
        -> adding search dir '/usr/libx32'
        Finished parsing ld.so.conf: '//etc/ld.so.conf.d/zz_x32-biarch-compat.conf'
    -> finished handling include directive '//etc/ld.so.conf.d/*.conf'
    Finished parsing ld.so.conf: '//etc/ld.so.conf'
Done scanning '/etc/ld.so.conf'
Search path:
-> '/lib'
-> '/usr/lib'
-> '/lib/aarch64-linux-gnu'
-> '/usr/lib/aarch64-linux-gnu'
-> '/lib/arm-linux-gnueabi'
-> '/usr/lib/arm-linux-gnueabi'
-> '/usr/local/lib'
-> '/lib/x86_64-linux-gnu'
-> '/usr/lib/x86_64-linux-gnu'
-> '/lib32'
-> '/usr/lib32'
-> '/libx32'
-> '/usr/libx32'
Scanning file 'deploy/usr/bin/my-qt-app'
    Parsing file 'deploy/usr/bin/my-qt-app'
    : search path:
    : - '/lib'
    : - '/usr/lib'
    : - '/lib/aarch64-linux-gnu'
    : - '/usr/lib/aarch64-linux-gnu'
    : - '/lib/arm-linux-gnueabi'
    : - '/usr/lib/arm-linux-gnueabi'
    : - '/usr/local/lib'
    : - '/lib/x86_64-linux-gnu'
    : - '/usr/lib/x86_64-linux-gnu'
    : - '/lib32'
    : - '/usr/lib32'
    : - '/libx32'
    : - '/usr/libx32'
    : end search path
    -> handling new dependency 'libQt5Network.so.5'
        Searching for 'libQt5Network.so.5'
        -> looking in '/lib' (root)
        -> looking in '/usr/lib' (root)
        -> looking in '/lib/aarch64-linux-gnu' (root)
        -> looking in '/usr/lib/aarch64-linux-gnu' (root)
            -> skip incompatible '//usr/lib/aarch64-linux-gnu/libQt5Network.so.5' (machine 'AArch64', need 'Advanced Micro Devices X86-64')
        -> looking in '/lib/arm-linux-gnueabi' (root)
        -> looking in '/usr/lib/arm-linux-gnueabi' (root)
            -> skip incompatible '//usr/lib/arm-linux-gnueabi/libQt5Network.so.5' (class 'ELF32', need 'ELF64')
        -> looking in '/usr/local/lib' (root)
        -> looking in '/lib/x86_64-linux-gnu' (root)
        -> looking in '/usr/lib/x86_64-linux-gnu' (root)
        ---> found
        libQt5Network.so.5 => /usr/lib/x86_64-linux-gnu/libQt5Network.so.5 (0xdeadbeef)
...
```

**Comment by nurupo** [comment]
The library calls `${triple}-readelf` on the binary to determine its Machine and Class, and checks those against the librarie's Machine and Class fiends:

```
$ readelf -Wh /usr/bin/gcc
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x403d00
  Start of program headers:          64 (bytes into file)
  Start of section headers:          1004320 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         10
  Size of section headers:           64 (bytes)
  Number of section headers:         33
  Section header string table index: 32
```

**Comment by nurupo** [comment]
@probonopd If you want this script included into linuxdeployqt, the script needs a user input on paths of the gcc and readelf binaries, as well as the path for -root flag, so linuxdeployqt would need to take this info from the user and pass on to the xldd. You can set `CT_XLDD_VERBOSE` env variable when calling xldd for `-verbose 3`.

**Comment by probonopd** [comment]
So you are saying it would __not__ be sufficient to rename _xldd_ to _ldd_ and put it inside the linuxdeployqt AppImage?

**Comment by nurupo** [comment]
Well, look at what the script does and try using it yourself. You will see that it requires some inputs. For example, there is no way for it to determine the sysroot of the toolchain you have used to build the binary. The sysroot could be anywhere on the system, it could be at `/home/probonopd/work/toolchains/test/that-one-nice-toolchain/`, or at `/tmp/tmp/tmp/toolchain` or somewhere else. How is ldd supposed to know where it is on your system and how to differentiate between seemingly the same `/home/probonopd/x86_64-linux-gnu` and `/home/probonopd/x86_64-linux-gnu-bak`?

**Comment by probonopd** [comment]
I see. So we need to a) bundle `xldd` and b) have an optional  `-sysroot` command line option. Do you think you could implement this and send a PR?

**Comment by nurupo** [comment]
Um, not, just sysroot is not enough.

>@probonopd If you want this script included into linuxdeployqt, the script needs a user input on paths of the gcc and readelf binaries, as well as the path for -root flag

**Comment by nurupo** [comment]
Huh. Looks like we have a couple of issues that need to be fixed before we can even considering using `xldd`.

My previous attempt to create an arm AppImage of my Qt app was stuck on linuxdeployqt not being able to run `ldd` on my app's executable, which is why I have opened this issue. Now I have tried to use the `xldd` with linuxdeployqt and I have hit another issue. I just noticed that linuxdeployqt runs `qmake` to get the sysroot path:

```
$ /usr/lib/x86_64-linux-gnu/qt5/bin/linuxdeployqt output/usr/share/applications/app.desktop -bundle-non-qt-libs -appimage
...
Log: Using qmake to determine the location of the Qt to be bundled
Log: Using qmake:  "/usr/lib/x86_64-linux-gnu/qt5/bin/qmake"
Log: -query output from qmake: "QT_SYSROOT:\nQT_INSTALL_PREFIX:/usr\nQT_INSTALL_ARCHDATA:/usr/lib/x86_64-linux-gnu/qt5\nQT_INSTALL_DATA:/usr/share/qt5\nQT_INSTALL_DOCS:/usr/share/qt5/doc\nQT_INSTALL_HEADERS:/usr/include/x86_64-linux-gnu/qt5\nQT_INSTALL_LIBS:/usr/lib/x86_64-linux-gnu\nQT_INSTALL_LIBEXECS:/usr/lib/x86_64-linux-gnu/qt5/libexec\nQT_INSTALL_BINS:/usr/lib/x86_64-linux-gnu/qt5/bin\nQT_INSTALL_TESTS:/usr/tests\nQT_INSTALL_PLUGINS:/usr/lib/x86_64-linux-gnu/qt5/plugins\nQT_INSTALL_IMPORTS:/usr/lib/x86_64-linux-gnu/qt5/imports\nQT_INSTALL_QML:/usr/lib/x86_64-linux-gnu/qt5/qml\nQT_INSTALL_TRANSLATIONS:/usr/share/qt5/translations\nQT_INSTALL_CONFIGURATION:/etc/xdg\nQT_INSTALL_EXAMPLES:/usr/lib/x86_64-linux-gnu/qt5/examples\nQT_INSTALL_DEMOS:/usr/lib/x86_64-linux-gnu/qt5/examples\nQT_HOST_PREFIX:/usr\nQT_HOST_DATA:/usr/lib/x86_64-linux-gnu/qt5\nQT_HOST_BINS:/usr/lib/x86_64-linux-gnu/qt5/bin\nQT_HOST_LIBS:/usr/lib/x86_64-linux-gnu\nQMAKE_SPEC:linux-g++-64\nQMAKE_XSPEC:linux-g++-64\nQMAKE_VERSION:3.0\nQT_VERSION:5.7.1\n"
Log: Qt libs path determined from qmake: "/usr/lib/x86_64-linux-gnu"
Log: Changed LD_LIBRARY_PATH: "/usr/lib/x86_64-linux-gnu:"
Log: Changed LD_LIBRARY_PATH: "/repo/build/output/usr/bin/../lib/:/usr/lib/x86_64-linux-gnu:"
Log: applicationBundle.libraryPaths: ("/repo/build/output/usr/lib/libQt5Network.so.5")
Log: additionalExecutables: ()
Log: allBinaryPaths: ("/repo/build/output/usr/bin/app", "/repo/build/output/usr/lib/libQt5Network.so.5")
Log: allRPaths: QSet("/usr/lib/x86_64-linux-gnu")
...
```

It finds the `/usr/lib/x86_64-linux-gnu/qt5/bin/qmake` and runs it, but the thing is that I'm creating an AppImage for `arm-linux-gnueabi`, so the things `/usr/lib/x86_64-linux-gnu/qt5/bin/qmake` binary would return wouldn't be correct. `/usr/lib/x86_64-linux-gnu/qt5/bin/qmake` would have returned correct things, but it's an arm executable, you can't run arm on x86_64 machine (I don't want to use qemu or anything like that), this is what you get when you try:

```
$ ARCH=armel /usr/lib/x86_64-linux-gnu/qt5/bin/linuxdeployqt output/usr/share/applications/app.desktop -bundle-non-qt-libs -appimage -qmake=/usr/lib/arm-linux-gnueabi/qt5/bin/qmake
...
ERROR: "/usr/lib/arm-linux-gnueabi/qt5/bin/qmake -query" exited with 2 : "/usr/lib/arm-linux-gnueabi/qt5/bin/qmake: 1: /usr/lib/arm-linux-gnueabi/qt5/bin/qmake: Syntax error: word unexpected (expecting \")\")\n"
ERROR: Qt path could not be determined from qmake on the $PATH
ERROR: Make sure you have the correct Qt on your $PATH
ERROR: You can check this with qmake -v
```

So... running the `qmake` is out of question, linuxdeployqt would need to be modified not to run qmake and get whatever input it needs from the user instead. Otherwise it's not usable for cross-compiling.

**Comment by nurupo** [comment]
Another issue is that linuxdeployqt tries to run the arm version of `appimagetool`, which results in

```
sh: 1: appimagetool: Exec format error
```

So appimagetool needs to have cross-compilation support too, unless it has it and I'm not using it right.

**Comment by nurupo** [comment]
@TheAssassin @probonopd alright guys, since there seem to be more things that need fixing than just `xldd`, some of which would require modifying linuxdeployqt and maybe AppImageKit too, I have setup [a minimal example of trying to cross-compile ARM AppImage on x86_64 system](https://github.com/nurupo/appimage-qt-cross-example), so that you could use it as a starting point when adding cross-compilation support to linuxdeployqt and AppImageKit. You can see the cross-compilation build script and Travis output when it runs it. It already uses `xldd`, it currently fails on linuxdeploqt trying to run ARM qmake on x86_64 system.

**Comment by probonopd** [comment]
Cross-compiling comes up from time to time as something users want, but I won't be working on it due to lack of time and other priorities. I simply don't understand enough about cross-compiling and usually run into issues whenever I try it. Hence, I am supportive of someone who wants to tackle this topic. To get started, please review https://github.com/AppImage/AppImageKit/search?q=cross-compile&unscoped_q=cross-compile&type=Issues.
